### PR TITLE
add route.Get() and skip PUT if route exists

### LINKED ISSUE #86300: Azure cloud provider shouldn't put route table when it's already been configured
**What happened**:

Kubernetes shouldn't put route table when it's already been configured. Azure cloud provider should check whether the route has already been created before invoking `az.CreateOrUpdateRoute(route)`.

https://github.com/kubernetes/kubernetes/blob/e680ad7156f263a6d8129cc0117fda58602e50ad/staging/src/k8s.io/legacy-cloud-providers/azure/azure_routes.go#L128-L131

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

/sig cloud-provider
/priority important-soon

### PR DESCRIPTION
**Pull request by zqingqing1**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Azure cloud provider shouldn't put route table when it's already been configured

**Which issue(s) this PR fixes**:

Fixes #86300 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Review by feiskyer — COMMENTED** [review]
/ok-to-test
/area provider/azure
/priority important-soon

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_routes.go [hunk 1]
```diff
@@ -184,6 +184,22 @@ func (az *Cloud) CreateRoute(ctx context.Context, clusterName string, nameHint s
 		},
 	}
 
+	actualRoute, err := az.RoutesClient.Get(ctx, az.RouteTableResourceGroup, az.RouteTableName, routeName)
```
Comment (feiskyer) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_routes.go]: could you switch to `az.ListRoutes()` here? `ListRoutes()` would get the routes from cache, which would help to reduce the chances of throttling issues.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_routes.go [hunk 2]
```diff
@@ -184,6 +184,22 @@ func (az *Cloud) CreateRoute(ctx context.Context, clusterName string, nameHint s
...
+	if err != nil {
+		klog.V(3).Infof("CreateRoute: creating route: failed(GetRoute) instance=%q with error=%v", kubeRoute.TargetNode, err)
+		return err
+	}
+
+	if actualRoute.RoutePropertiesFormat != nil &&
+		actualRoute.RoutePropertiesFormat.AddressPrefix != nil &&
+		*actualRoute.RoutePropertiesFormat.AddressPrefix == kubeRoute.DestinationCIDR &&
```
Comment (feiskyer) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_routes.go]: use `to.String()` is easier, and use `strings.EqualFold()` for case insentive, e.g.

e.g. `strings.EqualFold(to.String(actualRoute.RoutePropertiesFormat.AddressPrefix), kubeRoute.DestinationCIDR)`

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go [hunk 3]
```diff
@@ -113,6 +113,7 @@ type VirtualMachineScaleSetVMsClient interface {
 
 // RoutesClient defines needed functions for azure network.RoutesClient
 type RoutesClient interface {
+	Get(ctx context.Context, resourceGroupName string, routeTableName string, routeName string) (result network.Route, err error)
```
Comment (feiskyer) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go]: PR needs rebase. And please rename the returned `err error` to `rerr *retry.Error`, which is introduced in #86555.

Comment (zqingqing1) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go]: if we use az.ListRoutes(), then RoutesClient.Get() is no longer needed. I prefer not to include useless code, so I will remove this Get().

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_routes.go (hunk 1 above)
Comment (zqingqing1) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_routes.go]: sure, make more sense.

**Comment by zqingqing1** [comment]
I will squash commits before merging to the master after you confirm the PR is ready @feiskyer

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go (hunk 3 above)
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go]: Yep, it won't be required with LIST.

**Review by feiskyer — APPROVED** [review]
looks good, thanks. please squash the commits.

**Comment by zqingqing1** [comment]
squashed, should be good to be merged

**Review by feiskyer — COMMENTED** [review]
/lgtm
/approve

**Comment by feiskyer** [comment]
@zqingqing1 Thanks a lot for the fix

**Pull request merged**